Add qiskit.generic device that accepts instantiated qiskit backends.

airwoodix opened the pull request:
This patch introduces a new `qiskit.generic` device that accepts instantiated Qiskit backends and wraps them into a Pennylane device.

This offers a more composable mechanism for specifying the Qiskit backend, especially for providers that require more filter criteria than only a name to select a unique backend with [`get_backend`](https://qiskit.org/documentation/stubs/qiskit.providers.ProviderV1.get_backend.html#qiskit.providers.ProviderV1.get_backend).

Basic integration tests were added.

A small patch on the call to `JobV1.result()` was done: changes in #298 pass a `timeout` argument to [`JobV1.result()`](https://qiskit.org/documentation/stubs/qiskit.providers.JobV1.result.html#qiskit.providers.JobV1.result). While supported by many job handles, including that from IBMQ, it is *not* part of the ABC definition and is therefore not required to be valid.

timmysilv commented on pennylane_qiskit/generic.py:
```diff
@@ -0,0 +1,43 @@
+    short_name = "qiskit.generic"
+
+    def __init__(self, wires, backend, provider=None, shots=1024, **kwargs):
+        super().__init__(wires, provider=provider, backend=backend, shots=shots, **kwargs)
```
I see that you're trying to make provider optional, but is it possible to somehow match the parent class constructor? wouldn't need to copy-paste the docstring in that case

timmysilv commented on pennylane_qiskit/qiskit_device.py:
```diff
@@ -154,25 +162,26 @@ def __init__(self, wires, provider, backend, shots=1024, **kwargs):
 
-        self._backend = None
-
         self.provider = provider
```
I don't want to remove `provider` from `self` because users might depend on it, but it's not used outside of the `__init__` method anymore, so I kinda do...

timmysilv commented on pennylane_qiskit/qiskit_device.py:
```diff
@@ -154,25 +162,26 @@ def __init__(self, wires, provider, backend, shots=1024, **kwargs):
-        def _get_backend_name(name):
-            return name if isinstance(name, str) else name()
 
-        self._capabilities["backend"] = [
```
tbh I'm not entirely sure why this was added to the device capabilities, but again, users might depend on it. This PR could be considered "breaking" in that sense, but really this should have been private even in the old implementation

timmysilv reviewed (commented):
thanks for putting this together, it looks great! Device construction feels much more intuitive. just a few things to bring up:
- If you can test the V2 for `backend.name` and calling `job.result()` without timeout, it would be sweet. If not, we'll have to skip it somehow. Dealing with the changing IBMQ API can be tough sometimes so it's all good
- Can you give some details on the specific scenario in which you want to pass an already-existing device? It feels a bit funny making a new class just to implement `short_name` (not too bad, but I guess there are other ways around that if that's the only problem 😋 )

airwoodix commented:
Thanks for the positive review!

> If you can test the V2 for `backend.name` and calling `job.result()` without timeout, it would be sweet. If not, we'll have to skip it somehow. Dealing with the changing IBMQ API can be tough sometimes so it's all good

I'll have a look. It turns out to be difficult to find a concrete implementation of a `BackendV2` that doesn't also implement part of the `BackendV1` interface. Maybe a dummy would do as well?

> Can you give some details on the specific scenario in which you want to pass an already-existing device?

- if a backend needs to be configured (e.g. setting options), it's more convenient to do it before wrapping it as a PennyLane device that retrieving the backend through the private and untyped `self._capabilities["backend"]` value;
- some providers might require more than just a `name` string to uniquely identify a backend. This is allowed by [`ProviderV1.get_backend`](https://qiskit.org/documentation/stubs/qiskit.providers.ProviderV1.get_backend.html) using the extra keyword arguments. Rather than expanding the `qml.device()` signature to support this, I think it's easier and more elegant to use dependency injection.

> It feels a bit funny making a new class just to implement `short_name` (not too bad, but I guess there are other ways around that if that's the only problem yum )

I initially intended to use `QiskitDevice` directly. However, making the `provider` argument optional (null by default) would require modifying the argument order in `QiskitDevice.__init__`, which would be a major breaking change.

> 
> Don't forget to add a changelog entry!

Done. For the list of contributors, I'd be happy about adding an affiliation. Would that be fine for you?

timmysilv commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* Added a `GenericDevice` (PennyLane device name: `qiskit.generic`) that accepts a backend
```
slight name preference: could we call it `RemoteDevice` instead? Same with the rest of the changes.
```suggestion
* Added a `RemoteDevice` (PennyLane device name: `qiskit.remote`) that accepts a backend
```

timmysilv reviewed (commented):
this all makes sense, and looks good! Just one concern with naming. Also, could you update [index.rst](https://github.com/PennyLaneAI/pennylane-qiskit/blob/master/doc/index.rst) with a title card for this new device, and have it point to a new file in the `devices/` subfolder? It could be small like BasicAer's file, just something to show how/why to use this new device. Thanks!

timmysilv commented:
regarding your questions:
- don't worry much about the testing of V2 stuff, it's nearly identical
- those scenarios make sense. In the past, we'd have preferred you make a new device class for the specific device you're talking about, and having it accept the various custom args needed. This is an overall goal of the PennyLane plugins, but I can appreciate that this case goes beyond it (given the provider requirement). Perhaps we'll update the interaction between provider and backend in a later release.
- Changing the arg/kwarg order isn't the _best_ news, but it is essentially half of the point of this PR so it'll do. Plus, it's an isolated change
- the affiliation in the changelog is fine by us!

airwoodix commented:
Thanks for the review!

As requested, I renamed `qiskit.generic` to `qiskit.remote` and added a (very) short documentation in `doc/devices/remote.rst`.

I also added explicit `#pragma: no cover` directives for the backend v1/v2 and timeout/no timeout switches.

@timmysilv from your last comment, I understand that the remaining open threads do not need immediate action. Please let me know if that's not the case.

timmysilv commented on pennylane_qiskit/remote.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
```
```suggestion
# Copyright 2023 Xanadu Quantum Technologies Inc.
```

timmysilv reviewed (approved):
great stuff, thanks for addressing all our concerns! glad to have this in 🎉  and yes, you can disregard the previous comments 😄

airwoodix commented:
Sounds great, thanks a lot for considering this patch and reviewing it.

> re: docs failure

Thanks for the hint! Should be fixed now.

timmysilv commented:
looks great! I don't have permission to update the branch on your fork, but I fixed the codecov issue on master. If you pull in the latest master, I'll re-run CI and merge 😎

airwoodix commented:
Thanks! I rebased on latest `master`.
